quic/s2n-quic-core/src/buffer/reader/storage.rs

```diff
@@ -0,0 +1,75 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+mod buf;
+mod bytes;
+mod chunk;
+mod empty;
+mod full_copy;
+mod infallible;
+mod io_slice;
+mod slice;
+mod tracked;
+
+#[cfg(test)]
+mod tests;
+
+pub use buf::Buf;
+pub use chunk::Chunk;
+pub use empty::Empty;
+pub use full_copy::FullCopy;
+pub use infallible::Infallible;
+pub use io_slice::IoSlice;
+pub use tracked::Tracked;
+
+pub trait Storage {
+    type Error;
+
+    /// Returns the length of the chunk
+    fn buffered_len(&self) -> usize;
+
+    /// Returns if the chunk is empty
+    #[inline]
+    fn buffer_is_empty(&self) -> bool {
+        self.buffered_len() == 0
+    }
+
+    /// Reads the current contiguous chunk
+    fn read_chunk(&mut self, watermark: usize) -> Result<Chunk<'_>, Self::Error>;
+
+    /// Copies the reader into `dest`, with a trailing chunk of bytes.
+    ///
+    /// Implementations should either fill the `dest` completely or exhaust the buffered data.
+    ///
+    /// The storage may optionally return a `Chunk`, which can be used by the caller to defer
```

Comment:
```suggestion
    /// The storage also returns a `Chunk` for trailing data that didn't fit, which can be used by the caller to defer
```

Though it's not super clear to me what this chunk actually represents. The storage can have many non-contiguous chunks, right? So a single trailing chunk isn't necessarily trailing, it might be a "middle" chunk?

Are callers required to actually use the chunk (i.e., dropping/leaking it would lead to a missing chunk of data that's consumed by not written anywhere)?

Reply:
So the returned chunk must fit into the provided `dest`. It's just that we didn't copy it yet. So yes, the caller must use the `Chunk`, otherwise the data will never be written to the destination.

The returning chunk is specifically used for calling [LessSafeKey::seal_in_place_scatter](https://docs.rs/aws-lc-rs/1.6.1/aws_lc_rs/aead/struct.LessSafeKey.html#method.seal_in_place_scatter), where we can use the encryption as the copy.

Reply:
I've added some more comments here about how the caller is supposed to use chunk. Hopefully that's a bit better? But let me know if we need more there.

Reply:
Those comments look good to me.